### PR TITLE
3.x Minimal DefaultOutputTimezone support for Time/FrozenTime

### PR DESCRIPTION
**Pull request by ionas**
This PR tries to do the most minimal things that https://github.com/cakephp/cakephp/pull/8588 aimed for.

### CODE REVIEW

File: src/I18n/DateFormatTrait.php [hunk 1]
```diff
@@ -135,6 +170,11 @@ public function i18nFormat($format = null, $timezone = null, $locale = null)
     {
         $time = $this;
 
+        // This is required for testI18nFormatWithOffsetTimezone to pass
```
Comment (ionas) [inline src/I18n/DateFormatTrait.php]: I could need a hand with diffing natural e.g. 'America/Vancouver' and numerical e.g. '+09' DateTimeZones.

**Comment by ionas** [comment]
`static::$_isDateInstance` could be added thus that defaultOutputTimezone is only utilized by FrozenTime and Time not by Date if you want that.

File: tests/TestCase/I18n/TimeTest.php [hunk 2]
```diff
@@ -482,6 +490,40 @@ public function testI18nFormat($class)
...
+        $class::$defaultLocale = 'en-CA';
+        $class::setDefaultOutputTimezone('America/Vancouver');
+
+        $result = $time->i18nFormat();
+        $expected = '1/14/10 5:59 AM';
+        $this->assertTimeFormat($expected, $result);
+
+        $result = $time->i18nFormat(null, 'America/Toronto');
+        $expected = '1/14/10 8:59 AM';
+        $this->assertTimeFormat($expected, $result);
```
Comment (markstory) [inline tests/TestCase/I18n/TimeTest.php]: There should probably be tests for other methods like `nice()`, `__toString()` and `timeAgoInWords()`.

Comment (ionas) [inline tests/TestCase/I18n/TimeTest.php]: Added those.

**Comment by ionas** [comment]
@markstory I have added the missing tests. There are 3 tests failing and I can't figure out why. Can you take a look?

**Comment by ionas** [comment]
I have also tried to make it very clear that timezone conversion only works for time/datetime and never for date.

File: src/I18n/DateFormatTrait.php [hunk 3]
```diff
@@ -64,6 +75,39 @@
...
+        }
+        return static::$_defaultOutputTimezone;
+    }
+
+    /**
+     * Sets the default output timezone used by Time and FrozenTime.
+     *
+     * @param string|\DateTimeZone $timezone Timezone string or DateTimeZone object
+     * in which the date will be displayed.
+     * @return void
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php]: You'll need a `throws` doc tag to appease PHPCS.

File: src/I18n/DateFormatTrait.php [hunk 4]
```diff
@@ -157,9 +201,22 @@ public function i18nFormat($format = null, $timezone = null, $locale = null)
...
+        if ($time instanceof Time || $time instanceof FrozenTime) {
+            // Remove before merge:
+            // if ((is_string($time->timezone) && $time->timezone === 'UTC')
+            //     || ($time->timezone instanceof \DateTimeZone && $time->timezone->getName() === '+0:00')
+            // ) {
+            // }
+            // if ($time->getOffset() !== 0) {
+            //     $time = clone $time;
+            //     $time->add(new \DateInterval('PT' . $time->getOffset() / 60 . 'M'));
+            // }
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php]: What is this all about?

File: tests/TestCase/I18n/TimeTest.php (hunk 2 above)
Comment (markstory) [inline tests/TestCase/I18n/TimeTest.php]: Nice work on the additional tests.

File: src/I18n/DateFormatTrait.php (hunk 3 above)
Comment (ionas) [inline src/I18n/DateFormatTrait.php]: PHPCS seemed not to report that for me. I ll fix it.

File: src/I18n/DateFormatTrait.php (hunk 4 above)
Comment (ionas) [inline src/I18n/DateFormatTrait.php]: This is a dev note that's two different conditions I have been trying to get offsets + output time zones to work. If you take a look at the failing tests you will see the reason why. However I could not figure this last issue out.

**Comment by dereuromark** [comment]
:+1: Once the tests are green.
Awesome and important feature.

File: src/I18n/DateFormatTrait.php [hunk 5]
```diff
@@ -35,6 +38,14 @@
     public static $defaultLocale;
 
     /**
+     * The \DateTimeZone default output timezone used by Time and FrozenTime.
+     * See http://php.net/manual/en/timezones.php.
```
Comment (ravage84) [inline src/I18n/DateFormatTrait.php]: ``@see http....`` below ``@var``.
https://www.phpdoc.org/docs/latest/references/phpdoc/tags/see.html

File: tests/TestCase/I18n/TimeTest.php [hunk 6]
```diff
@@ -507,6 +603,33 @@ public function testI18nFormatWithOffsetTimezone($class)
     }
 
     /**
+     * test formatting dates with offset style timezone and defaultOutputTimezone
+     *
+     * @dataProvider classNameProvider
+     * @see https://github.com/facebook/hhvm/issues/3637
```
Comment (ravage84) [inline tests/TestCase/I18n/TimeTest.php]: Nitpicking, ``@see`` after return.

**Comment by davidyell** [comment]
I've had a quick scan through the diff and this looks great, it'll help lots of people as the issue of Storage v Display for datetimes with timezones seems to crop up a lot.

👍

**Comment by markstory** [comment]
I'm going to see if I can make headway on the tests for this.

File: tests/TestCase/I18n/TimeTest.php (hunk 6 above)
Comment (ionas) [inline tests/TestCase/I18n/TimeTest.php]: I hope you implied that @see should come after @return (alphabetically?)

**Comment by ionas** [comment]
@markstory fixed some docblocks thanks to @ravage84 - still the same tests :|

File: tests/TestCase/I18n/TimeTest.php (hunk 6 above)
Comment (ravage84) [inline tests/TestCase/I18n/TimeTest.php]: I think we never agreed on a certain order of those annotations, but personally I try to keep them in a kind of logical order.

1. parameter(s)
2. return value
3. throws
4. rest

For tests it's certainliy arguable whether ``@dataProvider`` can be seen as a parameter.
But ``@see`` clearly belongs to "the rest".

**Comment by markstory** [comment]
@ionas I had a chance to look at this last night. I'll push up a pull request later today.

**Comment by markstory** [comment]
Closing as a new pull request that incorporates these changes is open.

**Comment by ndm2** [comment]
@HJerem [**It has been reverted**](https://github.com/cakephp/cakephp/pull/8976) as there were unpleasant side-effects, the time helper's [**`outputTimezone` option**](https://github.com/cakephp/cakephp/pull/9003/files) is the alternative approach.

**Comment by HJerem** [comment]
@ndm2 Thanks for the answer, can you please give me an example on how to use it ? Can't find it in documentation.

**Comment by inoas** [comment]
For further support visit IRC (or Slack ☹)

* * *

If you look at the code @ndm2 linked you can see that it calls: https://secure.php.net/manual/de/function.date-default-timezone-get.php

Thus it will use defaults like `"Europe/Berlin"` ... so try setting something like that `"America/New_York"` ... see https://secure.php.net/manual/en/timezones.php ... https://secure.php.net/manual/en/timezones.europe.php etc.

You can try to configure the helper - without running code myself: `$this->Time->setConfig('outputTimezone', 'France/Paris')`

See https://github.com/cakephp/cakephp/blob/master/src/View/Helper/TimeHelper.php

**Pull request closed without merging**